src/api/app/models/review.rb

```diff
@@ -57,6 +64,13 @@ class NotFoundError < APIError
   before_validation :set_reviewable_association
   after_commit :update_cache
 
+  def review_assignment
+    errors.add(:unknown, 'no reviewer defined') unless by_user || by_group || by_project
+
+    return unless validate_reviewer_fields
```

Comment:
does it here make sense? 

Looking the possible values to make `validate_reviewer_fields` the possible combinations would return true, which looks wrong to me:

```
by_user: true by_group: false by_project: false by_package: true
by_user: false by_group: true by_project: true by_package: true
by_user: false by_group: true by_project: false by_package: true
by_user: true by_group: true by_project: false by_package: true
by_user: true by_group: true by_project: true by_package: true
by_user: true by_group: true by_project: false by_package: false
by_user: false by_group: true by_project: true by_package: false
by_user: true by_group: true by_project: true by_package: false
by_user: true by_group: false by_project: true by_package: true
by_user: true by_group: false by_project: true by_package: false
```

@coolo could you help me here?

Reply:
note that returning true is an error condition. So read the boolean expression as
```
  def validate_reviewer_fields
    return (by_group || by_project || by_package) if by_user
    return (by_project || by_package) if by_group
    false
  end
```

The function name implies that true is good, so that is the only problem I see :)

Reply:
hi @coolo, yes I didn't push my changes, sorry... I just need a new name for this method. Maybe `invalid_reviewers?`  

BTW I still think that the cases where the `validate_reviewer_fields` returns false, still don't match with the wanted logic right? 

I mean both

```
by_user: false by_group: false by_project: true by_package: true
by_user: false by_group: true by_project: false by_package: false
```

should be valid (and return false, but they returning true)

Reply:
by_package _requires_ by_project to be set as well.

Your 2nd line is only having by_group and as such is fine

Reply:
Basically the only valid options are:

```
by_user: true by_group: true by_project: false by_package: false
by_user: false by_group: true by_project: false by_package: false
by_user: false by_group: true by_project: true by_package: false
by_user: false by_group: true by_project: true by_package: true
```

Every other combination is invalid.

Reply:
Awesome, thank you!

Reply:
I had a bug in my initial comment that by_group was too often true - hope that didn't confuse you :)